src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,625 @@
+    }
+  }
+  if (build_state_ == kPass) {
+    try_load_locked();
+  }
+}
+
+NativeModule::~NativeModule() {
+  if (!is_global_ && dlopen_handle_) {
+    dlclose(dlopen_handle_);
+  }
+}
+
+bool NativeModule::try_wait_for_build_locked() {
+  if (build_state_ == kInit) {
+    // Since build or write-binary is performed synchronously while holding
+    // big_mutex, and we are holding big_mutex, we know that neither is 
+    // happening.  The lib_file may or may not exist: we just have to
+    // update the state of this NativeModule to match the filesystem.
+    build_state_ = (file_exists(lib_name_) ? kPass : kFail);
+    if (build_state_ == kFail) {
+      std::string base(config.module_dir_ + "/hm_" + key_);
+      fprintf(stderr, "makefile:\n%s", read_file_as_string(base+".mak").c_str());
+      fprintf(stderr, "errors:\n%s",   read_file_as_string(base+".err").c_str());
+    }
+  }
+  return (build_state_ == kPass);
+}
+
+bool NativeModule::try_load_locked() {
+  if (load_state_ == kInit) {
+    if (is_global_) {
+      load_state_ = kPass;
+    } else if (!try_wait_for_build_locked()) {
+      load_state_ = kFail;
+    } else {
+      auto handle = dlopen(lib_name_.c_str(), RTLD_GLOBAL|RTLD_NOW);
+      if (!handle) {
+        fprintf(stderr, "ERROR: dlopen failed: %s\n", dlerror());
+      }
+      load_state_ = (handle ? kPass : kFail);
+      if (handle) dlopen_handle_ = handle;
+    }
+  }
+  return (load_state_ == kPass);
+}
+
+std::vector<char> NativeModule::get_binary() {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  try_wait_for_build_locked();
+  int fd = open(config.get_lib_name(key_).c_str(), O_RDONLY, 0666);
+  std::vector<char> vec;
+  if (fd < 0) {
+    return vec;
+  }
+  struct stat st;
+  int rc = fstat(fd, &st);
+  assert(rc == 0);
+  size_t file_size = st.st_size;
+  vec.resize(file_size);
+  rc = read(fd, &vec[0], file_size);
+  assert(rc == (int)file_size);
+  close(fd);
+  return vec;
+}
+
+static std::string to_qualified_name(
+  JNIEnv* env,
+  const std::string& key,
+  jstring nameJ,
+  int numArgs,
+  bool is_global,
+  bool is_longfunc
+) {
+  JString name(env, nameJ);
+  char argTypeCodes[32];
+  for (int j = 0; j < numArgs; ++j) argTypeCodes[j] = 'l';
+  argTypeCodes[numArgs] = 0;
+  char buf[512];
+  if (is_global) {
+    // No name-mangling for global func names
+    strcpy(buf, name);
+  } else {
+    // Mangled name for hail::hm_<key>::funcname(NativeStatus* st, some numbers of longs)
+    auto moduleName = std::string("hm_") + key;
+    sprintf(buf, "_ZN4hail%lu%s%lu%sE%s%s",
+      moduleName.length(), moduleName.c_str(), strlen(name), (const char*)name, 
+      "P12NativeStatus", argTypeCodes);
+  }
+  return std::string(buf);
+}
+
+void NativeModule::find_LongFuncL(
+  JNIEnv* env,
+  NativeStatus* st,
+  jobject funcObj,
+  jstring nameJ,
+  int numArgs
+) {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  void* funcAddr = nullptr;
+  if (!try_load_locked()) {
```

Comment:
You try to load at the end of the worker constructor.  Should you do the same in the master constructor?  Then you don't need to lock here, and you can just check load_state.

Reply:
The usual lifecycle for the master module is that once we've got the key and binary, we close()
it without ever loading.  So eager loading would usually be redundant (though the cost is not necessarily noticeable).